keps/sig-api-machinery/3617-fine-grained-authz/README.md

```diff
@@ -0,0 +1,1121 @@
+* Whether the verb is excluded from or included in the generic privilege. This
+  is needed to achieve our second goal. Exclusion is expected to be very rare --
+  only new fields will be supported. Marking existing fields this way would
+  break existing clients unless the corresponding permission is added, which we
+  can't universally do (see authz section above).
+* If a field name changes between versions of an object, the permission verb
+  should not. Otherwise it will be extremely laborious to configure the system
+  to ensure permissions don't change when objects are accessed via different
+  versions.
+* Fields representing the same concept in *different* objects should share the
+  *same* permission. The authz system already has a spot for resource type,
+  which can be specific or wildcard based. We should not encode resource type
+  into the verb also.
+* Because the permissions are distinct from fields, we should do our best to
+  make the permissions (verb strings) not confusable for field names or paths.
+
+The exact permissions are going to be a shared responsibility between extension
+authors and in-tree Kubernetes developers, since anyone can author a CRD.
+
+As part of this KEP, we will update the schemas to supply permission verbs for
+the fields listed above. We will document the patterns used so that extension
+authors can make their own permission verbs which won't conflict with each other
+or with in-tree verbs. We may not know exactly what the best practices are until
+we have an alpha or beta behind us, hence we're not ready to specify that in
+this KEP yet.
+
+(The exact permission system will be described below because we're not done with
+the requirements yet, keep reading!)
+
+#### API Server Additional Plumbing and Checks
+
+Currently apiserver performs the authz check before descending into any detailed
+logic for mutating operations. This means that when it is doing the authz check,
+it knows the *object* that is being changed, but nothing about *which fields*
+are changing.
+
+We will introduce a new permission, "granular". This permission
+means that although the caller doesn't have generic update permissions, it may
+still change some fields if it has the specific permissions required.
+
+API Server already maintains schema information for all resource types. This
+will be watched to also maintain field <-> permission mappings.
+
+Now we can describe the modifications to API Server:
+
+On any mutating call,
+
+1. Check the "PUT" / "PATCH" / "CREATE" permission. If the actor has this,
+   proceed to step 3.
+2. Otherwise, check the "granular" permission. If yes, place a
+   marker in the request context and proceed to step 3, otherwise, fail the
+   request with a forbidden error.
+3. Compute the change (patch logic, SSA logic, defaulting etc). Compute a list
+   of fields which changed. (SSA logic makes this easy.)
+4. Pre- or post- (DECISION NEEDED) webhooks, check the list of fields:
```

Comment:
Probably post-mutating webhooks, no? I guess you could also run it both before and after.

Reply:
Kind of a trade off -- doing it post webhook is safer in terms of what gets committed but breaks webhooks that e.g. look at nodeName and keep a label up to date. pre-webhook is safer in terms of not exposing side-effect-having webhooks to malicious payloads.

Reply:
I prefer post-mutating webhooks as well.  I'd like to check on the actual object being stored.